tests/bq_test_helpers.py

```diff
@@ -0,0 +1,96 @@
+"""
+Shared helper functions for BigQuery-related tests
+"""
+import datetime
+import warnings
+from collections import OrderedDict
+from typing import Any, List, Union, Dict
+
+import mock
+from google.cloud import bigquery
+from google.cloud.bigquery import TableReference
+from google.cloud.bigquery.table import TableListItem
+
+_TYPE_TO_FIELD_TYPE = {
+    str(int): 'INT64',
+    str(str): 'STRING',
+    str(float): 'FLOAT64',
+    str(datetime.date): 'DATE',
+    str(datetime.datetime): 'DATETIME'
+}
+dict_types = Union[Dict, OrderedDict]
+
+
+def _field_from_key_value(key: str, value: Any) -> bigquery.SchemaField:
+    """
+    Get a schema field object from a key and value
+    
+    :param key: name of the field
+    :param value: value of the field
+    :return: an appropriate schema field object
+    """
+    tpe = str(type(value))
+    field_type = _TYPE_TO_FIELD_TYPE.get(tpe)
+    if not field_type:
+        raise NotImplementedError(f'The type for {value} is not supported')
+    return bigquery.SchemaField(name=key, field_type=field_type)
+
+
+def fields_from_dict(row):
+    """
+    Get schema fields from a row represented as a dictionary
+
+    :param row: the dictionary to infer schema for
+    :return: list of schema field objects
+    
+    Example:
+        >>> from tests import bq_test_helpers
+        >>> d = {'item': 'book', 'qty': 2, 'price': 1.99}
+        >>> bq_test_helpers.fields_from_dict(d)
+            [SchemaField('item', 'STRING', 'NULLABLE', None, ()),
+             SchemaField('qty', 'INT64', 'NULLABLE', None, ()),
+             SchemaField('price', 'FLOAT64', 'NULLABLE', None, ())]
+    """
+    return [_field_from_key_value(key, value) for key, value in row.items()]
+
+
+def mock_query_result(rows: List[dict_types]):
+    """
+    Create a mock RowIterator as returned by :meth:`bigquery.QueryJob.result`
+    from rows represented as a list of dictionaries
+
+    :param rows: a list of dictionaries representing result rows
+    :return: a mock RowIterator
+    """
+    mock_row_iter = mock.MagicMock(spec=bigquery.table.RowIterator)
+    mock_row_iter.total_rows = len(rows)
+    row0 = rows[0]
+    if isinstance(row0, OrderedDict):
+        _rows = rows
+    else:
+        warnings.warn('Provide rows as `OrderedDict` if type conversion '
+                      'results in unexpected behavior.')
```

Comment:
So any time I use this without providing an ordered dictionary, it's going to fuss at me?

Reply:
Yeah, I suspect there may be edge cases where passing a list of plain dicts yields inconsistent row structure. If this is true it would likely be difficult to troubleshoot the issue. I thought adding this to the function docstring might be TL;DR, but we can do that if you think it's a better option.